### ISSUE TITLE
Feature request: allow specifying files to exclude in ArchiveFiles task

### ISSUE THREAD

**Issue by midgleyc** [microsoft/azure-pipelines-tasks#9272]
## Feature Request

Currently, ArchiveFiles adds every file in a given directory to the created zip file. It would be nice to be able to specify particular files/folders to not publish even if present, e.g. ".git". This can be worked around by adding a script that moves unwanted files outside the directory before running the ArchiveFiles task, but it would be nicer to be able to do this as part of the task itself.

**Comment by midgleyc** [comment]
If I could use an equivalent of the "contents:" option from CopyFiles (a newline separated list of files to include, with excluded files beginning with !), I think that would be ideal.

**Comment by leantk** [comment]
This issue is stale because it has been open for a year with no activity. Remove the stale label or comment on the issue otherwise this will be closed in 5 days

**Comment by asdkant** [comment]
a built-in option to ignore file and folder names that usually are not included in a release (.git folder, dev-stage-only stuff from the language, intermediary files from compilation, etc.) would also be useful in keeping down the size of final files and the time it takes to compress them.

**Comment by robatwilliams** [comment]
This is still relevant.

The symmetry of having the `Contents` option like the sibling CopyFiles and DeleteFiles tasks would be good.

**Comment by klodoma** [comment]
+1

**Comment by klodoma** [comment]
Any workaround to this one? Or do I have to remove everything we don't want before creating the archive?

**Comment by LiamKarlMitchell** [comment]
Instead of ArchiveFiles task found I could use Publish Pipeline Artifacts with a .artifactignore file.
Which let me deploy it in a release build.

```yml
          - task: PublishPipelineArtifact@1
            inputs:
              targetPath: $(rootFolder)
              artifact: WebApp
              publishLocation: 'pipeline'
```

Did the trick.

**Comment by 9jaboy** [comment]
@LiamKarlMitchell is the Artifact compressed or zipped?

**Comment by LiamKarlMitchell** [comment]
Not aware.
https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/utility/publish-pipeline-artifact?view=azure-devops

**Comment by 9jaboy** [comment]
ok, it won't be a substitute for most.

**Comment by liliankasem** [comment]
+1 it would be great to get this feature implemented! 

In the meantime, I used the CopyFiles task to create a folder with the items I needed, then used the ArchiveFiles task to create a zip of that folder.

**Comment by BlakesHeaven** [comment]
This seems like such an important and obvious option to include I can't believe it wasn't in the original scope of requirements.
There are always going to be some files/folder we don't want to include, eg *.sqlproj or other development scripts not intended for others.

In my particular case, the zip being created is used by a DevOps Release pipeline which has a stage to copy to a fileshare for manual deployment, possibly by a customer! Sure we can export zip to temp folder, delete unwanted stuff, re-zip and delete temp folder before copying, but it would be so much easier to simply exclude files/folders in the first place.

It is now half way through 2021, 3.5 years since this was raised; is this ever likely to see the light of day?
Can we have an update please?

**Comment by tbeavers** [comment]
Plain and simple, being able to exclude specific files/folders when using ArchiveFiles is a much needed option. In the meantime, will be using an additional CopyFiles task to "stage" what I actually want to archive.

**Comment by yuryleb** [comment]
Anybody home? This option is really necessary. Or just excluding 'hidden' files (started with dot) will work too.

**Comment by dexmoocat** [comment]
+1

**Comment by lesscodetxm** [comment]
> ok, it won't be a substitute for most.

Yes, I have a similar issue where I want to create an artifact out of a subset of the files in a certain folder, and I want the artifact to be a zip file. Using an `.artifactignore` file and `PublishPipelineArtifact` is perfect to do the filtering, but in order to get it as a zip file I'm having to do a dance involving `DownloadPipelineArtifact` on the first artifact, `ArchiveFiles` on that, and then `PublishPipelineArtifact` again on the .zip. This results in two redundant artifacts, since I can't (easily) remove one of them, nor can I replace the original (unzipped) artifact.

I know this directly address the OP's issue, but having an option to zip on `PublishPipelineArtifact` would simplify this use case tremendously.

Or maybe I'm missing a simpler way that already exists?

**Comment by D-Angelov** [comment]
+1

Excluding files from an archive is what most people do, it would be really beneficial if it can be added to ArchiveFiles tasks.

**Comment by baannee** [comment]
+1

**Comment by alanmk** [comment]
In some cases, removing the unnecessary files with the `DeleteFiles@1` task might serve as a workaround.

**Comment by jcarasick-geog8r** [comment]
still needed

**Comment by thomeu** [comment]
+1

**Comment by bfprenticeb** [comment]
+1

Adding a folder strictly to include/exclude specific files is unnecessary overhead.

**Comment by yoHasse** [comment]
+1

**Comment by Baldassari** [comment]
+1

**Comment by PL-EduardoSanchez** [comment]
+1

**Comment by cirio** [comment]
Any news about this feature? It's would be very useful

**Comment by robertoprog** [comment]
+1

**Comment by peter-bozovic** [comment]
+1

**Comment by anastasiia-petrenko** [comment]
+1

**Comment by midgleyc** [comment]
If you're posting +1, make sure to also react with 👍 to the first post -- I understand that's a metric that might actually get used for prioritisation, whereas the comments just stave off stalebot.

**Comment by klodoma** [comment]
Go with the workarounds.... this is 3 years old... I don't see this getting through.

**Comment by denisyakimov07** [comment]
+1

**Comment by Takadonet** [comment]
+1

**Comment by matt-lethargic** [comment]
+1 seems like a basic requirement and not that much work

**Comment by denisyakimov07** [comment]
+1

**Comment by tosagartambe** [comment]
I used this task before archiving files. Hope this helps for time being.

- task: Bash@3
  inputs:
    targetType: 'inline'
    script: |
      echo '*******************rsync*******************'
      rsync -av --exclude='.git' --exclude='app/Config/app.config . code

**Comment by lizdeika** [comment]
Been strugling but then realized I could just bash and keep calm :D
```yaml
- bash: zip -r ${{ variables.ARTIFACT_NAME }} . -x .git/\*
      displayName: 'Create deployment artifact'
```

**Comment by Vicky050890** [comment]
Hi @anilsakr,  
Your suggestion worked for me. Thanks

**Comment by ewencodes** [comment]
+ 1

**Comment by robatwilliams** [comment]
We still want it please, just there isn't really to discuss hence the lack of activity.

**Comment by ppanchal-mobiz** [comment]
Bump! Reacted to the post and commented to avoid this becoming stale. Much needed.

**Comment by netdragonboberb** [comment]
I assume you should use DeleteFiles@1  https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/delete-files-v1?view=azure-pipelines
We have a 4GB .git dir so we have to do this too.

**Comment by robatwilliams** [comment]
We still want it please.

This ask isn't going to go stale.

**Comment by tudor33sud** [comment]
+1 !

**Comment by hudsy13** [comment]
+1

**Comment by robatwilliams** [comment]
> We still want it please.
>
> This ask isn't going to go stale.

**Comment by hudsy13** [comment]
Hard to understand why this is such a hard request to fulfill
